lib/ansible/module_utils/basic.py

```diff
@@ -2602,33 +2606,31 @@ def run_command(self, args, check_rc=False, close_fds=True, executable=None, dat
             strings on python3, use encoding=None to turn decoding to text off.
         '''
 
-        if isinstance(args, list):
-            if use_unsafe_shell:
-                args = " ".join([shlex_quote(x) for x in args])
-                shell = True
-        elif isinstance(args, (binary_type, text_type)):
-            if use_unsafe_shell:
-                shell = True
-            else:
-                # On python2.6 and below, shlex has problems with text type
-                # On python3, shlex needs a text type.
-                if PY2:
-                    args = to_bytes(args, errors='surrogate_or_strict')
-                elif PY3:
-                    args = to_text(args, errors='surrogateescape')
-                args = shlex.split(args)
-        else:
+        if isinstance(args, (binary_type, text_type)):
+            # On python2.6 and below, shlex has problems with text type
+            # On python3, shlex needs a text type.
+            if PY2:
+                args = to_bytes(args, errors='surrogate_or_strict')
+            elif PY3:
+                args = to_text(args, errors='surrogateescape')
+            args = shlex.split(args)
+        elif not isinstance(args, list):
             msg = "Argument 'args' to run_command must be list or string"
             self.fail_json(rc=257, cmd=args, msg=msg)
 
+        # expand shellisms
+        args = [os.path.expanduser(os.path.expandvars(x)) for x in args if x is not None]
+
         shell = False
         if use_unsafe_shell:
+
+            # not set explicitly, check if set by controller
             if executable is None:
-                executable = os.environ.get('SHELL')
-            if executable:
-                args = [executable, '-c', args]
-            else:
-                shell = True
+                if self._shell not in (None, '/bin/sh'):
+                    args = [self._shell, '-c'] + args
```

Comment:
This seems wrong.  Won't this end up being the equivalent of:
```
/bin/sh -c if [ x"test" = x"test" ] ; then printf "hi" ; fi
```

When what we really want is the former which is the equivalent of:
```
 /bin/sh -c 'if [ x"test" = x"test" ] ; then printf "hi" ; fi' 
```

Reply:
updated to always stringify